File: homunculy/src/internal/domain/repositories/agent_repository.py
"""Agent repository interface."""

from abc import ABC, abstractmethod
from typing import List, Optional

from internal.domain.entities import Agent, AgentConfiguration


class AgentRepository(ABC):
    """Agent persistence contract."""

    @abstractmethod
    async def create(self, agent: Agent) -> Agent:
        """Create a new agent."""
        raise NotImplementedError

    @abstractmethod
    async def get_by_id(self, agent_id: str) -> Optional[Agent]:
        """Get agent by id."""
        raise NotImplementedError

    @abstractmethod
    async def update(self, agent: Agent) -> Agent:
        """Update an agent."""
        raise NotImplementedError

    @abstractmethod
    async def delete(self, agent_id: str) -> bool:
        """Delete an agent by id."""
        raise NotImplementedError

    @abstractmethod
    async def list_all(self, limit: int = 50, offset: int = 0) -> List[Agent]:
        """List agents with pagination."""
        raise NotImplementedError

    @abstractmethod
    async def update_configuration(self, agent_id: str, configuration: AgentConfiguration) -> bool:
        """Update agent configuration."""
        raise NotImplementedError

File: management-service/internal/domain/repositories/agent_repository.go
package repositories

import (
	"context"
	"management-service/internal/domain/entities"
)

// AgentRepository defines the interface for agent data operations
type AgentRepository interface {
	// Create saves a new agent to the repository
	Create(ctx context.Context, agent *entities.Agent) error

	// GetByID retrieves an agent by its ID
	GetByID(ctx context.Context, id string) (*entities.Agent, error)

	// GetByName retrieves an agent by its name
	GetByName(ctx context.Context, name string) (*entities.Agent, error)

	// List retrieves all agents with optional filtering
	List(ctx context.Context, limit, offset int) ([]*entities.Agent, error)

	// Update modifies an existing agent
	Update(ctx context.Context, agent *entities.Agent) error

	// Delete removes an agent by ID
	Delete(ctx context.Context, id string) error

	// Exists checks if an agent with the given ID exists
	Exists(ctx context.Context, id string) (bool, error)

	// Count returns the total number of agents
	Count(ctx context.Context) (int64, error)
}
